### PR TITLE
Refactor sudt with registry address

### PR DESCRIPTION
**Pull request by classicalliu**

### CODE REVIEW

**Review by Flouse — APPROVED** [review]

File: packages/api-server/src/methods/modules/gw.ts [hunk 1]
```diff
@@ -347,26 +347,42 @@ export class Gw {
 
   /**
    *
-   * @param args [short_script_hash]
+   * @param args [scriptHash(Hash), registryId(HexNumber)]
    * @returns
    */
-  async get_script_hash_by_short_script_hash(args: any[]) {
+  async get_registry_address_by_script_hash(args: any[]) {
```
Comment (RetricSu) [inline packages/api-server/src/methods/modules/gw.ts:353]: consider adding a simple cache like `get_script_hash_by_registry_address` later?

**Review by RetricSu — APPROVED** [review]

File: crates/indexer/src/helper.rs [hunk 2]
```diff
@@ -117,18 +83,47 @@ pub enum GwLog {
     },
 }
 
-fn parse_sudt_log_data(data: &[u8]) -> ([u8; 20], [u8; 20], u128) {
-    assert_eq!(data[0], 20);
-    let mut from_address = [0u8; 20];
-    from_address.copy_from_slice(&data[1..21]);
+fn parse_sudt_log_data(data: &[u8]) -> anyhow::Result<(RegistryAddress, RegistryAddress, u128)> {
+    let mut start = 0;
+    let mut end = 0;
```
Comment (keroro520) [inline crates/indexer/src/helper.rs:88]: assert!(data.len() >= 8);

File: crates/indexer/src/helper.rs [hunk 3]
```diff
@@ -117,18 +83,47 @@ pub enum GwLog {
...
 
-    let mut to_address = [0u8; 20];
-    to_address.copy_from_slice(&data[21..41]);
+    end += {
+        let from_address_byte_size = u32::from_le_bytes(data[4..8].try_into()?);
+        if from_address_byte_size == 0 {
+            8
+        } else {
+            28
```
Comment (keroro520) [inline crates/indexer/src/helper.rs:95]: I am confused about this assignment.
Would it be possible to add some comments to the code: what's our expected format of `data`?

File: packages/godwoken/src/types.ts [hunk 4]
```diff
@@ -54,3 +55,8 @@ export interface AccountMerkleState {
   merkle_root: Hash;
   count: HexU32;
 }
+
+export interface RegistryAddress {
+  registry_id: HexNumber;
```
Comment (keroro520) [inline packages/godwoken/src/types.ts:60]: I think `HexU32` is more accurate than `HexNumber`.

File: packages/godwoken/src/client.ts [hunk 5]
```diff
@@ -42,24 +43,36 @@ export class GodwokenClient {
...
+    const registryAddress = await this.rpcCall(
+      "get_registry_address_by_script_hash",
+      scriptHash,
+      toHex(registryId)
+    );
+    return registryAddress;
+  }
+
+  public async getScriptHashByRegistryAddress(
+    registryAddress: HexString
```
Comment (keroro520) [inline packages/godwoken/src/client.ts:59]: Is it possible to use [`RegistryAddress`](https://github.com/nervosnetwork/godwoken-web3/pull/282/files#diff-a1a6f34c884302fbf9b7dd717268a3fa8bbf49b1950af04c2e90cd9e028172c2R59) as the parameter type? "HexString" is a bit ambiguous.

File: packages/godwoken/src/client.ts [hunk 6]
```diff
@@ -42,24 +43,36 @@ export class GodwokenClient {
...
+    const scriptHash = await this.rpcCall(
+      "get_script_hash_by_registry_address",
+      registryAddress
     );
     return scriptHash;
   }
 
   public async getBalance(
-    shortScriptHash: HexString,
+    registryAddress: HexString,
```
Comment (keroro520) [inline packages/godwoken/src/client.ts:69]: ditto

**Review by keroro520 — APPROVED** [review]

File: crates/indexer/src/helper.rs (hunk 2 above)
Comment (classicalliu) [reply crates/indexer/src/helper.rs]: Actually, data length has been checked before enter this function

File: crates/indexer/src/helper.rs (hunk 3 above)
Comment (classicalliu) [reply crates/indexer/src/helper.rs:98]: Add some comments of this function

File: packages/godwoken/src/types.ts (hunk 4 above)
Comment (classicalliu) [reply packages/godwoken/src/types.ts]: updated

File: packages/godwoken/src/client.ts (hunk 5 above)
Comment (classicalliu) [reply packages/godwoken/src/client.ts]: Renamed to `serializedRegistryAddress`

**Pull request merged**